test/AutoDiff/adjoint_expr_silgen.swift

```diff
@@ -0,0 +1,210 @@
+// RUN: %target-swift-frontend -emit-sil -verify %s | %FileCheck %s
+
+// Note: the adjoint calculations in this file are not intended to be accurate.
+
+//===----------------------------------------------------------------------===//
+// Top-level function
+//===----------------------------------------------------------------------===//
+
+@differentiable(reverse, adjoint: dSquare(_:primal:seed:))
+func square<T : FloatingPoint>(_ x: T) -> T {
+  return x * x
+}
+@_silgen_name("dsquare")
+private func dSquare<T : FloatingPoint>(_ x: T, primal: T, seed: T) -> T {
+  return 2 * x
+}
+
+@_silgen_name("test_top_level")
+func testTopLevel() -> Float {
+  let adjointSquare: (Float, Float, Float) -> Float = #adjoint(square)
+  return adjointSquare(1, 2, 3)
+}
+// CHECK-LABEL: sil hidden @test_top_level
+// CHECK: [[DSQUARE_RAW:%.*]] = function_ref @dsquare : $@convention(thin) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> @out τ_0_0
```

Comment:
do we care about the details of " (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @in_guaranteed τ_0_0) -> @out τ_0_0" in this test? if not, it'd be good to omit them, to avoid unnecessary test maintenance work in the future.

e.g. in case @in_guaranteed is updated to some other calling convention.

Reply:
I agree that the extra test maintenance is unpleasant, but I would like to keep the function signatures in the test because they're informative.

Many function_refs are (partially) applied and it's difficult to tell what's happening without seeing the function type. For instance methods, it's especially important to verify that `#adjoint` has the intended type (it should be curried, first taking a `self` instance argument).

If the default calling convention changes frequently, I'd be happy to refactor this test.

Reply:
how does the SIL code here help verify that it takes a self instance argument? it seems that info will need to be in the next line: on partial_apply

Reply:
How do we read the function_ref inst to determine it's being curried or not?

Reply:
Here's an toy example adapted from `adjoint_expr_silgen.swift`:
```swift
// Original Swift.
extension Pair {
  @differentiable(reverse, ...)
  func subtract(_ a: Pair) -> Pair { ... }

  private func dSubtract(_ a: Pair, primal: Pair, seed: Pair) -> (Pair, Pair) { ... }
}

@_silgen_name("test_pair")
func testPair() {
  _ = #adjoint(Pair.subtract)
}
```
```llvm
// Output from `swiftc -emit-silgen`.
// test_pair
sil hidden @test_pair : $@convention(thin) () -> () {
bb0:
  // function_ref curry thunk of Pair.dSubtract(_:primal:seed:)
  %0 = function_ref @$xxx : $@convention(thin) (Pair) -> @owned @callee_guaranteed (Pair, Pair, Pair) -> (Pair, Pair) // user: %1
  ... // omitted
} // end sil function 'test_pair'
```

The type of the method `xxx` is (ignoring calling convention markers) `(Pair) -> (Pair, Pair, Pair) -> (Pair, Pair)`. This is currying, like the example you provided above.

Normally in Swift, when applying an instance method like `x.dSubtract(...)`, the `self` instance (in this case `x`) is applied implicitly.
However, if you reference an instance method directly using `Pair.dSubtract`, its type will be curried (because there's no `self` instance).

For `#adjoint` expression, we decided to make `#adjoint(<instance method>)` always have a curried type because that's more general and flexible.

Reply:
Thanks! This is very helpful.